x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockFactory.java

```diff
@@ -0,0 +1,265 @@
+            BlockFactoryParameters.class,
+            BlockFactory.class.getClassLoader()
+        );
+        List<ServiceLoader.Provider<BlockFactoryParameters>> impls = loader.stream().toList();
+        if (impls.size() != 1) {
+            throw new AssertionError("expected exactly one impl, but got:" + impls);
+        }
+        BlockFactoryParameters params = impls.get(0).get();
+        return new BlockFactory(params.breaker(), params.bigArrays());
+    }
+
+    private final CircuitBreaker breaker;
+
+    private final BigArrays bigArrays;
+
+    private final boolean checkBreaker = true;
+
+    private BlockFactory(CircuitBreaker breaker, BigArrays bigArrays) {
+        this.breaker = breaker;
+        this.bigArrays = bigArrays;
+    }
+
+    /**
+     * Returns the global ESQL block factory.
+     */
+    public static BlockFactory getGlobalInstance() {
+        return GLOBAL;
+    }
+
+    /**
+     * Returns the Non-Breaking block factory.
+     */
+    public static BlockFactory getNonBreakingInstance() {
+        return NON_BREAKING;
+    }
+
+    public static BlockFactory getInstance(CircuitBreaker breaker, BigArrays bigArrays) {
+        return new BlockFactory(breaker, bigArrays);
+    }
+
+    // For testing
+    public CircuitBreaker breaker() {
+        return breaker;
+    }
+
+    // For testing
+    public BigArrays bigArrays() {
+        return bigArrays;
+    }
+
+    /**
+     * Adjust the circuit breaker with the given delta, if the delta is
+     * negative, or checkBreaker is false, the breaker will be adjusted
+     * without tripping.  If the data was already created before calling
+     * this method, and the breaker trips, we add the delta without breaking
+     * to account for the created data.  If the data has not been created yet,
+     * we do not add the delta to the breaker if it trips.
+     */
+    void adjustBreaker(final long delta, final boolean isDataAlreadyCreated) {
+        if (checkBreaker) {
+            // checking breaker means potentially tripping, but it doesn't
+            // have to if the delta is negative
+            if (delta > 0) {
+                try {
+                    breaker.addEstimateBytesAndMaybeBreak(delta, "<esql_block_factory>");
+                } catch (CircuitBreakingException e) {
+                    if (isDataAlreadyCreated) {
+                        // since we've already created the data, we need to
+                        // add it so closing the stream re-adjusts properly
+                        breaker.addWithoutBreaking(delta);
+                    }
+                    // re-throw the original exception
+                    throw e;
+                }
+            } else {
+                breaker.addWithoutBreaking(delta);
+            }
+        } else {
+            // even if we are not checking the breaker, we need to adjust
+            // its' totals, so add without breaking
+            breaker.addWithoutBreaking(delta);
+        }
+    }
```

Comment:
This is pretty reasonable. I kind of wish we didn't need the `isAlreadyCreated` bit, but we're sort of stuck with it. There are far worse things though.

Reply:
👍 Yeah, we kinda need the flexibility to be able to create blocks in different ways.